File: day1/cpp/calibrate2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

//COULD BE FURTHER OPTIMIZED... SEARCHING FOR LITERAL AND STRING VERSION OF NUMBERS IN EACH LINE
void checkPositions(int& l, int& r, string line)
{
    int pos, n = line.size(), min = n, max = -1;
    vector<string> vals = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

    for(int i=0;i<vals.size();i++) {
        //SEARCHING FROM FRONT OF STRING FOR MIN
        pos = line.find(vals[i]);

        if (pos != string::npos && pos < min) {
            min = pos;
            l = i%9 + 1;
        }
        //SEARCHING FROM BACK OF STRING FOR MAX
        pos = line.rfind(vals[i]);
        if (pos != string::npos && pos > max) {
            max = pos;
            r = i%9 + 1;
        }
    }
}

int sumCalibrations()
{
    //OPENING FILE WHEN DECLARING IFSTREAM
    ifstream f("../input.txt");
    string line;
    int n, tmp, start, end, sum;

    sum = 0;

    //READING FILE INPUT LINE BY LINE

    while (f.is_open() && getline(f, line)) {

        start = end = -1;

        checkPositions(start, end, line);
        
        // ADD 2 DIGIT NUMBER FORMED BY START & END TO SUM
        sum += start * 10 + end;
    }

    f.close();

    return sum;
}

int main()
{
    int sum = sumCalibrations();
    cout << "SUM OF CALIBRATION CODES: " << sum << endl;
    return 0;
}

File: day1/rust/calibrate2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let mut sum = 0;

    if let Ok(lines) = read_lines("../input.txt") {

        for line in lines {
            if let Ok(ip) = line {
                sum += check_numbers(ip);
            }
        }

        println!("Final sum: {}", sum);
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn check_numbers(s: String) -> i32 {
    let numstrings = vec!["1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    let n = s.len() as usize;
    let mut l: usize = 0;
    let mut r: usize = 0;
    let mut min: usize = n;
    let mut max: usize = 0;

    for (i, num) in numstrings.iter().enumerate() {
        if let Some(pos) = s.rfind(num) {
            if pos >= max {
                r = i%9+ 1;
                max = pos;
            }
        }
        if let Some(pos) = s.find(num) {
            if pos <= min {
                l = i%9 + 1;
                min = pos;
            }
        }

    }
    
    let sum = (l * 10 + r) as i32;

    sum
}